### PR TITLE
Implementation of invalid user input exception

### PR DESCRIPTION
**Pull request by NlightNFotis**
This PR contains an implementation of an erroneous user input exception, and changes the unstructured `throw 0`s in the `cbmc/` folder into structured exceptions and preconditions/invariants.

### CODE REVIEW

File: src/cbmc/cbmc_solvers.cpp [hunk 1]
```diff
@@ -237,20 +241,42 @@ std::unique_ptr<cbmc_solverst::solvert> cbmc_solverst::get_smt2(
...
+      "");
+  }
+  else if(incremental_check)
   {
-    error() << "sorry, this solver does not support incremental solving" << eom;
-    throw 0;
+    throw invalid_user_input_exceptiont(
+      "Reason: sorry, this solver does not support incremental solving",
+      "--incremental-check",
+      "");
```
Comment (hannes-steffenhagen-diffblue) [inline src/cbmc/cbmc_solvers.cpp]: This is quite a big improvement over the previous version that didn't tell you what option it didn't like, :+1:

File: src/util/exception_utils.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    : reason(reason), option(option), correct_input(correct_input)
+  {
+  }
+
+  std::string what() const noexcept
+  {
+    std::string res;
+    res += "\nInvalid User Input Exception\n";
+    res += "Option: " + option + "\n";
+    res += reason;
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/exception_utils.h]: I think it's fine to have the `"Reason: "` part here, I just didn't like forcing the form of the part after the reason.
Sorry for being unclear with this

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
LGTM for the most part

File: src/cbmc/bv_cbmc.cpp [hunk 3]
```diff
@@ -9,16 +9,12 @@ Author: Daniel Kroening, kroening@kroening.com
...
 
 bvt bv_cbmct::convert_waitfor(const exprt &expr)
 {
-  if(expr.operands().size()!=4)
-  {
-    error().source_location=expr.find_source_location();
-    error() << "waitfor expected to have four operands" << eom;
-    throw 0;
-  }
+  PRECONDITION(expr.operands().size() != 4);
```
Comment (danpoe) [inline src/cbmc/bv_cbmc.cpp]: We could use `DATA_INVARIANT` here which is for checking whether goto programs, exprts, etc. are well-formed. Then we can also keep the error message.

File: src/cbmc/show_vcc.cpp [hunk 4]
```diff
@@ -143,7 +144,8 @@ void bmct::show_vcc()
   {
     of.open(filename);
     if(!of)
-      throw "failed to open file "+filename;
+      throw invalid_user_input_exceptiont(
+        "Failed to open suggested file", "--outfile", "");
```
Comment (danpoe) [inline src/cbmc/show_vcc.cpp]: We can drop "suggested", and append the filename to the first argument.

File: src/util/exception_utils.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+  std::string reason;
+  std::string option;
+  std::string correct_input;
+
+public:
+  invalid_user_input_exceptiont(
+    std::string reason,
+    std::string option,
+    std::string correct_input)
```
Comment (danpoe) [inline src/util/exception_utils.h]: Can we have an empty string default argument for `correct_input`?

File: src/util/exception_utils.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  std::string what() const noexcept
+  {
+    std::string res;
+    res += "\nInvalid User Input Exception\n";
+    res += "Option: " + option + "\n";
+    res += "Reason: " + reason;
+    // Print an optional correct usage message assuming correct input parameters have been passed
+    correct_input.empty() ? "\n" : res += " Try: " + correct_input + "\n";
```
Comment (danpoe) [inline src/util/exception_utils.h]: I think we need to add a newline before `" Try"`.

File: src/cbmc/cbmc_main.cpp [hunk 7]
```diff
@@ -52,4 +58,9 @@ int main(int argc, const char **argv)
   #endif
 
   return res;
+  }
+  catch(invalid_user_input_exceptiont &e)
```
Comment (danpoe) [inline src/cbmc/cbmc_main.cpp]: Currently if an exception is caught here the program still exits with a zero exit code. We should change this to exit with an error code.

I think we can also move the exception catching code to `parse_options_baset::main()` (wrapping the call to `doit()`). Then the tools that inherit from it don't need to repeat the exception handlers.

**Review by danpoe — CHANGES_REQUESTED** [review]

File: src/util/exception_utils.h (hunk 6 above)
Comment (NlightNFotis) [reply src/util/exception_utils.h]: I'd rather it be kept on the same line.

File: src/cbmc/bv_cbmc.cpp (hunk 3 above)
Comment (NlightNFotis) [reply src/cbmc/bv_cbmc.cpp]: I won't implement this. It's easily implementable, but the error message is just redundant IMO. It doesn't convey any information that the precondition expression doesn't convey already. It just duplicates semantic information about the behaviour of the code.

File: src/cbmc/cbmc_main.cpp (hunk 7 above)
Comment (NlightNFotis) [reply src/cbmc/cbmc_main.cpp]: This is very interesting. It's a great suggestion, but this needs a little bit thinking about how to properly implement this (I haven't thought it through yet, might be easier than I realise)

File: src/cbmc/bv_cbmc.cpp (hunk 3 above)
Comment (danpoe) [reply src/cbmc/bv_cbmc.cpp]: Hm the error message had the additional information that `expr` is a wait-for expression (i.e., `expr.id()` is `"waitfor"`). From the condition we only see that it is some expression with four operands.

Comment (NlightNFotis) [reply src/cbmc/bv_cbmc.cpp]: Good spot, didn't notice that. I guess I can rewrite that a little bit as well.

File: src/util/exception_utils.h (hunk 6 above)
Comment (danpoe) [reply src/util/exception_utils.h]: Should we add a separator character then? Maybe a comma.

**Comment by NlightNFotis** [comment]
@hannes-steffenhagen-diffblue @danpoe Can I request a new review from both of you? Made some substantial changes on where we handle exceptions in the latest commit.

File: src/cbmc/cbmc_main.cpp [hunk 8]
```diff
@@ -25,6 +25,10 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <iostream>
 #endif
 
+#include <iostream>
+
+#include <util/exception_utils.h>
```
Comment (danpoe) [inline src/cbmc/cbmc_main.cpp]: I think those two includes are now not needed anymore.

File: src/cbmc/show_vcc.cpp [hunk 9]
```diff
@@ -143,7 +144,8 @@ void bmct::show_vcc()
   {
     of.open(filename);
     if(!of)
-      throw "failed to open file "+filename;
+      throw invalid_user_input_exceptiont(
+        "Failed to open file" + filename, "--outfile");
```
Comment (danpoe) [inline src/cbmc/show_vcc.cpp]: `... file"` -> `...file "`

**Review by danpoe — APPROVED** [review]
Looks good to me!

File: src/cbmc/show_vcc.cpp [hunk 10]
```diff
@@ -143,7 +144,8 @@ void bmct::show_vcc()
   {
     of.open(filename);
     if(!of)
-      throw "failed to open file "+filename;
+      throw invalid_user_input_exceptiont(
+        "Failed to open file: " + filename, "--outfile");
```
Comment (hannes-steffenhagen-diffblue) [inline src/cbmc/show_vcc.cpp:148]: Thinking about it, I don't really like this being an invalid user input exception (I suppose you can justify it being subject to user input as they control the filename, but technically whether or not we're able to open the file isn't necessarily up to the user).

Not a blocker for sure, but in general I'd have a preference for IO related exceptions to have their own place.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
Still looks OK to me

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**